metricbeat/module/system/process/network/tracker.go

```diff
@@ -0,0 +1,259 @@
+	TCP uint64
+	UDP uint64
+}
+
+// CounterUpdateEvent is sent every time we get new packet data for a PID
+type CounterUpdateEvent struct {
+	pktLen        int
+	TransProtocol applayer.Transport
+	Proc          *common.ProcessTuple
+}
+
+// RequestCounters is a request for packet data
+type RequestCounters struct {
+	Pid  int
+	Resp chan PacketData
+}
+
+// Tracker tracks network packets and maps them to a PID
+type Tracker struct {
+	procData    map[int]PacketData
+	dataMut     sync.RWMutex
+	procWatcher *procs.ProcessesWatcher
+
+	log    *logp.Logger
+	gctime time.Duration
+
+	updateChan chan CounterUpdateEvent
+	reqChan    chan RequestCounters
+	stopChan   chan struct{}
+
+	// special test helpers
+	loopWaiter chan struct{}
+	testmode   bool
+}
+
+// NewNetworkTracker creates a new network tracker for the given config
+func NewNetworkTracker() (*Tracker, error) {
+	watcher := &procs.ProcessesWatcher{}
+	err := watcher.Init(procs.ProcsConfig{Enabled: true})
+	if err != nil {
+		return nil, fmt.Errorf("error initializing process watcher: %w", err)
+	}
+
+	tracker := &Tracker{
+		procData: map[int]PacketData{},
+		dataMut:  sync.RWMutex{},
+
+		updateChan: make(chan CounterUpdateEvent, 10),
+		reqChan:    make(chan RequestCounters, 10),
+		stopChan:   make(chan struct{}, 1),
+		gctime:     time.Minute * 10,
+		// right now, the packetbeat watcher won't work with alternate mountpoints,
+		// as support is missing from go-sysinfo. This means we can't support /hostfs settings
+		procWatcher: watcher,
+
+		log: logp.L(),
+	}
+
+	return tracker, nil
+}
+
+// Update the tracker with the given counts
+func (track *Tracker) Update(packetLen int, proto applayer.Transport, proc *common.ProcessTuple) {
+	track.updateChan <- CounterUpdateEvent{pktLen: packetLen, TransProtocol: proto, Proc: proc}
+}
+
+// Get data for a given PID
+func (track *Tracker) Get(pid int) PacketData {
+	req := RequestCounters{Pid: pid, Resp: make(chan PacketData)}
+	track.reqChan <- req
+	got := <-req.Resp
+	return got
+}
+
+// Stop the tracker
+func (track *Tracker) Stop() {
+	track.stopChan <- struct{}{}
+}
+
+// Track is a non-blocking operation that starts a packet sniffer, and the underlying
+// tracker that correlates packet data with pids
+func (track *Tracker) Track(ctx context.Context) error {
+	helperContext, cancel := context.WithCancel(ctx)
+	go func() {
+		if !track.testmode {
+			err := StartPacketHandle(helperContext, track.procWatcher, track)
+			if err != nil {
+				track.log.Errorf("error starting packet capture: %s", err)
+				ctx.Done()
+			}
+		}
+	}()
+
+	go func() {
+		track.garbageCollect(helperContext)
+	}()
+
+	go func() {
+		for {
+			select {
+			case update := <-track.updateChan:
+				track.updateInternalTracking(update)
+			case req := <-track.reqChan:
+				track.dataMut.RLock()
+				if proc, ok := track.procData[req.Pid]; ok {
+					req.Resp <- proc
+				} else {
+					req.Resp <- PacketData{}
+				}
+				track.dataMut.RUnlock()
```

Comment:
This can block for a long time while writing to the channels, better to unlock immediately.

```suggestion
				proc, ok := track.procData[req.Pid]
				track.dataMut.RUnlock()
				if ok {
					req.Resp <- proc
				} else {
					req.Resp <- PacketData{}
				}
```